python/src/boombox/endpoints.py

```diff
@@ -0,0 +1,395 @@
+"""Boombox endpoints.
+
+Endpoints are classes defining possible inputs and outputs of
+:py:class:`.Boombox`. Each endpoint represents a different media format and has
+appropriate attributes that describe it.
+
+Examples:
+  * MP4("path/to/file.mp4") - an endpoint defining an MP4 container. If
+    provided for input, then Boombox will read a MP4 file from this location. If
+    provided for output, then Boombox will create a file at that location and
+    store the produced stream in it in MP4 format.
+  * HLS("path/to/playlist") - an endpoint defining a HLS playlist. Only
+    output is supported. If used, a playlist will be created in the specified
+    location.
+  * WebRTC("ws://host:port") - an endpoint defining a WebRTC connection.
+    Both input and output is supported. Websocket at the provided URL is used as
+    a signaling channel.
+
+"""
+
+from dataclasses import dataclass, KW_ONLY, fields, is_dataclass
+from term import Atom
+from typing import Any, Literal
+from typing_extensions import override
+from abc import ABC
+
+
+@dataclass
+class BoomboxEndpoint(ABC):
+    """Abstract base class of a Boombox endpoint.
+
+    Boombox endpoints are the definitions of Boombox inputs and outputs. This
+    class is a base class for these definitions. When creating a Boombox
+    instance it expects a specification of it's input and output as
+    BoomboxEndpoints.
+
+    Attributes
+    ----------
+    transcoding_policy : {None, "if_needed", "always", "never"}
+        Allowed only for output. The default transcoding behavior is "if_needed",
+        which means that if the format of the media is the same for input and
+        output, then the stream is not decoded and encoded. This approach saves
+        resources and time, but in some cases transcoding can be necessary. To
+        force transcoding regardless if the formats differ or not, this option
+        can be set to "always". If set to "never", Boombox will never transcode,
+        raising if the desired operation can't be done without transcoding.
+    """
+
+    _: KW_ONLY
+    transcoding_policy: Literal["if_needed", "always", "never"] | None = None
+
+    def get_atom_fields(self) -> set[str]:
+        """:meta private:"""
+        return {"transcoding_policy"}
```

Comment:
How about iterating through your class `fields` and filter only these fields that have `field.kw_only`?